testdata/cheats/File.t.sol

```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: Unlicense
+pragma solidity >=0.8.0;
+
+import "ds-test/test.sol";
+import "./Cheats.sol";
+
+contract FileTest is DSTest {
+    Cheats constant cheats = Cheats(HEVM_ADDRESS);
+
+    function testReadFile() public {
+        string memory path = "../testdata/fixtures/File/read.txt";
+
+        assertEq(cheats.readFile(path), "hello readable world\nthis is the second line!");
+    }
+
+    function testReadLine() public {
+        string memory path = "../testdata/fixtures/File/read.txt";
+
+        assertEq(cheats.readLine(path), "hello readable world");
+        assertEq(cheats.readLine(path), "this is the second line!");
+        assertEq(cheats.readLine(path), "");
+    }
+
+    function testWriteFile() public {
+        string memory path = "../testdata/fixtures/File/write_file.txt";
+        string memory data = "hello writable world";
+        cheats.writeFile(path, data);
+
+        assertEq(cheats.readFile(path), data);
+
+        cheats.removeFile(path);
+    }
+
+    function testWriteLine() public {
```

Comment:
How do these work in the real world? I.e. if I have two tests that write to the same file, will it work as expected?

I'm asking because one of primary use cases (IIRC) for writing to files was to record inputs and stuff in fuzz tests, so it's conceivable that this would be written to the same file in a format like `<test name>,<output>` or something.

Reply:
it seems to me that it'll work the same way as writing to the same file from Rust? So if it's a POSIX-compliant system, most of the writes (especially small ones like lines) should be atomic and not interleavable, see for example https://stackoverflow.com/a/1154599

Reply:
yeh, also all fuzz cases for the _same_ test are called sequentially.
and even if there's another test that reads7writes the same file we can expect them to be atomic I think

Reply:
> I'm asking because one of primary use cases (IIRC) for writing to files was to record inputs and stuff in fuzz tests, so it's conceivable that this would be written to the same file in a format like `<test name>,<output>` or something.

Yep, you're thinking of #1022 (cc @transmissions11). Since runs of a fuzz test execute serially like @mattsse said, it seems we could call `writeLine` in the fuzz test to update the output file in each run, and that is arguably sufficient to close that issue. Two downsides though:
- You wouldn't be able to write a header row since we can't identify which fuzzer run we're in. Possible workaround is to add a cheatcode to return the current fuzz run number?
- Writing to the file in each fuzz run will be a lot slower than one batch write at the end.

We might also want to add a `vm.toString()` or `vm.stringify()` cheatcode which converts value types to a string. There are solidity libraries that help with this, but none are perfect (e.g. leading zeros are truncated if you stringify addresses, since they pass through uint)

Reply:
>You wouldn't be able to write a header row since we can't identify which fuzzer run we're in. Possible workaround is to add a cheatcode to return the current fuzz run number?

Arguably you could also write this header in `setUp` :)

Reply:
Ahh interesting, nice workaround! I don't love that it fragments file writing, e.g. some of the code for it is done in `setUp` and some in the test itself, but definitely ok for now

Reply:
> Writing to the file in each fuzz run will be a lot slower than one batch write at the end.

this is my primary concern

ffi is incredibly slow right now— hoping this much better

Reply:
yeah it should be much faster

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

import "forge-std/Test.sol";

contract ContractTest is Test {
    function setUp() public {}

    function _writeToFile(string memory path, string memory line) internal {
        string[] memory inputs = new string[](5);
        inputs[0] = "/Users/shekhirin/Projects/bench-forge-write/write-to-file.sh";
        inputs[1] = "-f";
        inputs[2] = path;
        inputs[3] = "-i";
        inputs[4] = line;

        vm.ffi(inputs);
    }

    function testFFI() public {
        for (uint i = 0; i < 1000; i++) {
            _writeToFile("ffi.txt", "gm");
        }

        vm.removeFile("ffi.txt");
    }

    function testWriteLine() public {
	for (uint i = 0; i < 1000; i++) {
            vm.writeLine("writeLine.txt", "gm");
	}

	vm.removeFile("writeLine.txt");
    }
}
```

```console
➜  bench-forge-write git:(main) ✗ ../foundry/target/release/forge test --ffi --match-test testFFI
[⠢] Compiling...
No files changed, compilation skipped

Running 1 test for test/Contract.t.sol:ContractTest
[PASS] testFFI() (gas: 4789029)
Test result: ok. 1 passed; 0 failed; finished in 1.81s

➜  bench-forge-write git:(main) ✗ ../foundry/target/release/forge test --match-test testWriteLine
[⠢] Compiling...
No files changed, compilation skipped

Running 1 test for test/Contract.t.sol:ContractTest
[PASS] testWriteLine() (gas: 650236)
Test result: ok. 1 passed; 0 failed; finished in 23.69ms
```

Reply:
yeah, looks like a general `std::process::Command` overhead
<img width="1551" alt="image" src="https://user-images.githubusercontent.com/5773434/174820139-a833ab41-8c54-4ae5-9ee0-589725dfc43a.png">

Reply:
awesome